File: Храмцов/2192.py
class Solution(object):
    def getAncestors(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[List[int]]
        """
        ancestors = [set() for i in range(n)]

        new_edges = [[] for i in range(n)]

        for edge in edges:
            ancestors[edge[1]].add(edge[0])
            new_edges[edge[0]].append(edge[1])

        queue = []

        for i in range(n):
            if (len(ancestors[i]) == 0):
                queue.append(i)

        while len(queue) > 0:
            u = queue.pop(0)
            print(u)
            for v in new_edges[u]:
                ancestors[v] = ancestors[u].union(ancestors[v])
                if (v not in queue):
                    queue.append(v)

        for i in range(n):
            ancestors[i] = list(ancestors[i])
            ancestors[i].sort()
        return ancestors

File: Пукаев/2192.cpp
class Solution {
public:

    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector<vector<int>> result(n);
        vector<vector<int>> graphVector(n);
       
        for(auto edge: edges){
            graphVector[edge[0]].push_back(edge[1]);
        }
        for(int i = 0; i < n; i++){
            vector<bool> findedVertex(n, false);
            DFS(graphVector, result, i, i, findedVertex);
        }
        return result;

    }

    void BFS(int n, vector<vector<int>>& graphVector, int v, vector<vector<int>>& result){
        int startV = v;
        vector<bool> findedVertex(n, false);
        queue<int> q;
        
        q.push(v);
        findedVertex[v] = true;

        while (!q.empty()){
            v = q.front();
            q.pop();
            
            for (auto u: graphVector[v]){
                if (!findedVertex[u]){
                    findedVertex[u] = true;
                    q.push(u);
                    result[u].push_back(startV);
                }
            }
        }
    }

    void DFS(vector<vector<int>>& graphVector, vector<vector<int>>& result, int v, int u,vector<bool>& findedVertex){
        findedVertex[u] = true;
        for(auto vert: graphVector[u]){
            if(!findedVertex[vert]){
                result[vert].push_back(v);
                DFS(graphVector, result, v, vert, findedVertex);
            }
        }
    }






   
};
